### ISSUE TITLE
[torchx/slurm scheduler] Retries and retry policies

### ISSUE THREAD

**Issue by kiukchung** [pytorch/torchx#394]
## Description
Support:

(1) job restarts when the job is preempted - do we have support for this? I think it depends on the SLURM setup on the cluster and iirc FAIR cluster started doing this automatically. Normally, there are infinite resumptions upon preemption

(2) the ability to set job max retries upon timeout

see - https://github.com/facebookresearch/pycls/blob/8c79a8e2adfffa7cae3a88aace28ef45e52aa7e5/pycls/core/distributed.py#L116

## Motivation/Background
For FAIR's omnivision on SLURM usecase

## Alternatives
N/A

## Additional context/links
N/A

**Comment by mannatsingh** [comment]
From https://github.com/facebookincubator/submitit/blob/6f9e1f67178b08b050576fe6bc02e4555568128a/docs/checkpointing.md - 

> When you want to train your model, you just have to run the following code, and it will be submitted to a slurm job, which will be checkpointed and requeued at most `slurm_max_num_timeout=3` times if timed out (and any number of time if preempted)

> On Slurm cluster, you can trigger a fake preemption or timeout in order to test your checkpointing by using the job method: job._interrupt(timeout=<False,True>).

So for preemptions we just need to make sure that by default torchx jobs requeue upon pre-emption or take that as an arg from the user.

**Comment by d4l3k** [comment]
It seems like submitit handles this via a runtime but I think we might be able to do it via the sbatch control script. We'll have to play with it more. Adding a runtime has a fair number of complications around where to save the # of retries remaining. Doesn't seem like there's anything as simple as "--num-retries " flag for sbatch/srun

**Comment by rmadamson** [comment]
> It seems like submitit handles this via a runtime but I think we might be able to do it via the sbatch control script. We'll have to play with it more. Adding a runtime has a fair number of complications around where to save the # of retries remaining. Doesn't seem like there's anything as simple as "--num-retries " flag for sbatch/srun

Would something like the sbatch "--requeue" flag do what you want?

**Comment by mannatsingh** [comment]
Ah I see - I think this isn't an immediate blocker. It's only applicable when we want to jobs which are expected to run longer than the max timeout (set by admins). 
I think submitit saves its state in a pickle file in the launch directory and that's where this information probably also gets saved.
